### PR TITLE
[MOL-9398][BX] added image render for form-select

### PR DESCRIPTION
**Pull request by wilsontbx**
**Changes**
added image render for form select

- delete branch

**Changelog entry**

- [BREAKING | WARNING | remove if none] Description of change in `Component-name`

**Additional information**

- You may refer to this [MOL-9398](https://jira.ship.gov.sg/browse/MOL-9398)
- Any other information you would like to include

### CODE REVIEW

**Review by qroll — COMMENTED** [review]
hmm can we generalise this component to take in a custom `renderListItem` instead?

File: src/input-select/types.ts [hunk 1]
```diff
@@ -60,6 +60,8 @@ export interface InputSelectProps<T, V>
     displayValueExtractor?: (option: T) => string | undefined;
     /** Function to convert value into a string */
     valueToStringFunction?: (value: V) => string | undefined;
+    /** Function to render selected custom component */
+    renderCustomSelectedItem?: (item: string | V) => JSX.Element | undefined;
```
Comment (qroll) [inline src/input-select/types.ts:64]: per above comment, amend typing

also rename to be consistent with the other props 

```suggestion
    renderCustomSelectedOption?: (item: T) => JSX.Element | undefined;
```

File: src/input-select/input-select.tsx [hunk 2]
```diff
@@ -157,7 +159,9 @@ export const InputSelect = <T, V>({
                 </PlaceholderLabel>
             );
         } else {
-            return (
+            return renderCustomSelectedItem ? (
```
Comment (qroll) [inline src/input-select/input-select.tsx:162]: can move this nested ternary into the if-else?

i.e.
```
if (!selected) {
  // stuff
} else if (renderCustomSelectedItem) {
  // more stuff
} else {
  // stuff
}
```

File: src/input-select/input-select.tsx [hunk 3]
```diff
@@ -157,7 +159,9 @@ export const InputSelect = <T, V>({
...
+                renderCustomSelectedItem(getDisplayValue())
```
Comment (qroll) [inline src/input-select/input-select.tsx:163]: I think passing in the original option would provide the most flexibility, if other properties are needed for rendering

File: src/util/string-helper.ts [hunk 4]
```diff
@@ -86,4 +86,8 @@ export namespace StringHelper {
 
         return text;
     };
+
+    export const isImageUrl = (url: string) => {
```
Comment (qroll) [inline src/util/string-helper.ts:90]: can trouble you to revert this change if it's not needed for this PR?

File: src/shared/dropdown-list/types.ts [hunk 5]
```diff
@@ -6,6 +6,8 @@ export interface DropdownDisplayExtractorProps<T, V> {
     valueExtractor?: (item: T) => V | undefined;
     /** Function to derive options display value from an item */
     listExtractor?: (item: T) => string | undefined;
+    /** Function to render custom component */
+    renderListItem?: (item: T) => JSX.Element | undefined;
```
Comment (qroll) [inline src/shared/dropdown-list/types.ts:10]: might want to rename `DropdownDisplayExtractorProps` to `DropdownDisplayProps` because this interface is no longer strictly for extractors

File: src/shared/dropdown-list/dropdown-list.tsx [hunk 6]
```diff
@@ -295,12 +297,16 @@ export const DropdownList = <T, V>({
...
                                 />
                             )}
-                            <Label truncateType={itemTruncationType}>
-                                {itemTruncationType === "middle" &&
-                                hasExceededContainer(item)
-                                    ? renderTruncatedText(item)
-                                    : getOptionLabel(item)}
-                            </Label>
+                            {renderListItem ? (
+                                renderListItem(item)
```
Comment (qroll) [inline src/shared/dropdown-list/dropdown-list.tsx:301]: `ListItemSelector` has a fixed 3.5rem height. can you run this past UX to see if future dropdowns might need larger, dynamically sized items? can perhaps relax the fixed height if `renderListItem` is defined

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/shared/dropdown-list/dropdown-list.tsx (hunk 6 above)
Comment (wilsontbx) [reply src/shared/dropdown-list/dropdown-list.tsx:301]: Will check with UX tmr

File: stories/form/form-select/form-select.stories.mdx [hunk 7]
```diff
@@ -72,6 +73,54 @@ import { Form } from "@lifesg/react-design-system/form";
     </Story>
 </Canvas>
 
+<Heading3>With custom display</Heading3>
```
Comment (keithtxw) [inline stories/form/form-select/form-select.stories.mdx:76]: Rename to `With custom list item display`

File: stories/form/form-select/form-select.stories.mdx [hunk 8]
```diff
@@ -72,6 +73,54 @@ import { Form } from "@lifesg/react-design-system/form";
...
+
+You are able to customise the rendering of the selected option and the dropdown items.
```
Comment (keithtxw) [inline stories/form/form-select/form-select.stories.mdx:78]: **Edit**:
You are able to customise the display of the dropdown list items as well as their selected state.

File: stories/form/form-select/form-select.stories.mdx [hunk 9]
```diff
@@ -72,6 +73,54 @@ import { Form } from "@lifesg/react-design-system/form";
...
+
+<Canvas>
+    <Story name="With custom display">
```
Comment (keithtxw) [inline stories/form/form-select/form-select.stories.mdx:81]: Use "With custom list item display"

File: stories/form/form-select/form-select.stories.mdx [hunk 10]
```diff
@@ -72,6 +73,54 @@ import { Form } from "@lifesg/react-design-system/form";
...
+        <StoryContainer>
+            <Container>
+                <Form.Select
+                    label="This has custom rendering"
```
Comment (keithtxw) [inline stories/form/form-select/form-select.stories.mdx:85]: Use "This has a custom list item display"

File: stories/form/form-select/props-table.tsx [hunk 11]
```diff
@@ -110,6 +110,18 @@ const DATA: ApiTableSectionProps[] = [
...
+            {
+                name: "renderListItem",
+                description:
+                    "Function to customise the rendering of a list item",
+                propTypes: ["(option: T) => JSX.Element"],
+            },
+            {
+                name: "renderCustomSelectedOption",
+                description:
+                    "Function to customise the rendering of the selected option object",
```
Comment (keithtxw) [inline stories/form/form-select/props-table.tsx:122]: Update description to:
"Function to customise the rendering of a list item's selected state"

File: stories/form/form-select/doc-element.tsx [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import styled from "styled-components";
+
+export const ImageWrapper = styled.div`
+    display: flex;
+    border-radius: 4px;
+    border: 1px solid;
```
Comment (keithtxw) [inline stories/form/form-select/doc-element.tsx:6]: is there no color? if there isnt a color, can just specify `border-width`

**Review by keithtxw — CHANGES_REQUESTED** [review]

File: stories/form/form-select/doc-element.tsx (hunk 12 above)
Comment (qroll) [reply stories/form/form-select/doc-element.tsx:6]: found that minimally need to specify `border-style`

color will inherit the current color

File: stories/form/form-select/doc-element.tsx [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import styled from "styled-components";
```
Comment (keithtxw) [inline stories/form/form-select/doc-element.tsx:1]: rename file to `doc-elements.tsx`

**Review by keithtxw — CHANGES_REQUESTED** [review]

**Review by qroll — APPROVED** [review]

**Review by keithtxw — APPROVED** [review]
